src/lib/storage/lz4/lz4_encoder.hpp

```diff
@@ -0,0 +1,125 @@
+#include "storage/lz4_segment.hpp"
+#include "storage/value_segment.hpp"
+#include "storage/value_segment/value_segment_iterable.hpp"
+#include "storage/vector_compression/vector_compression.hpp"
+#include "types.hpp"
+#include "utils/assert.hpp"
+#include "utils/enum_constant.hpp"
+
+#include "lib/lz4hc.h"
+
+namespace opossum {
+
+class LZ4Encoder : public SegmentEncoder<LZ4Encoder> {
+ public:
+  static constexpr auto _encoding_type = enum_c<EncodingType, EncodingType::LZ4>;
+  static constexpr auto _uses_vector_compression = false;
+
+  template <typename T>
+  std::shared_ptr<BaseEncodedSegment> _on_encode(const std::shared_ptr<const ValueSegment<T>>& value_segment) {
+    const auto alloc = value_segment->values().get_allocator();
+    const auto num_elements = value_segment->size();
+    DebugAssert(num_elements <= std::numeric_limits<int>::max(),
+                "Trying to compress a ValueSegment with more "
+                "elements than fit into an int.");
+
+    // TODO(anyone): when value segments switch to using pmr_vectors, the data can be copied directly instead of
+    // copying it element by element
+    auto values = pmr_vector<T>{alloc};
+    values.reserve(num_elements);
+    auto null_values = pmr_vector<bool>{alloc};
+    null_values.reserve(num_elements);
+
+    // copy values and null flags from value segment
+    auto iterable = ValueSegmentIterable<T>{*value_segment};
+    iterable.with_iterators([&](auto it, auto end) {
+      for (; it != end; ++it) {
+        auto segment_value = *it;
+        values.emplace_back(segment_value.value());
+        null_values.push_back(segment_value.is_null());
+      }
+    });
+
+    // lz4 compression
+    const auto input_size = static_cast<int>(values.size() * sizeof(T));
+    auto output_size = LZ4_compressBound(input_size);
+    auto compressed_data = pmr_vector<char>{alloc};
+    compressed_data.reserve(static_cast<size_t>(output_size));
+    const int compression_result = LZ4_compress_HC(reinterpret_cast<char*>(values.data()), compressed_data.data(),
+                                                   input_size, output_size, LZ4HC_CLEVEL_MAX);
+    if (compression_result <= 0) {
+      Fail("LZ4 compression failed");
+    }
+
+    auto data_ptr = std::allocate_shared<pmr_vector<char>>(alloc, std::move(compressed_data));
+    auto null_values_ptr = std::allocate_shared<pmr_vector<bool>>(alloc, std::move(null_values));
+
+    return std::allocate_shared<LZ4Segment<T>>(alloc, data_ptr, null_values_ptr, nullptr, compression_result,
+                                               input_size, num_elements);
+  }
+
+  std::shared_ptr<BaseEncodedSegment> _on_encode(
+      const std::shared_ptr<const ValueSegment<std::string>>& value_segment) {
+    const auto alloc = value_segment->values().get_allocator();
+    const auto num_elements = value_segment->size();
+
+    // copy values and null flags from value segment
+    auto values = pmr_vector<char>{alloc};
```

Comment:
Have you tested reserving for values as well (By summing over the lengths of all strings in the segment in the beginning)?

Reply:
No we haven't benchmarked the performance difference but now that you mention it, we will do so 👍

Reply:
We found that the difference is negligible ([benchmark](http://quick-bench.com/aplSNPOV_-7cKL6sP_gNhFqBtKg)) but we can add it still add it (since there is a small performance improvement).

Reply:
resize + []= is even faster than reserving if you know the length in advance and the default constructor is trivial.

Reply:
Yes, but I don't think replacing multiple characters at once is possible (in a vector of characters). At least I didn't find a way to do that. And iterating also over the single characters in the strings is very slow ([benchmark](http://quick-bench.com/ARtSFQTuCxt8fnsgQWhVoJvZkfA)).

Reply:
ok

Reply:
If you still want to look into this, `std::copy` instead of `v.insert` should do the job.

Reply:
`std::copy` seems to be [super slow](http://quick-bench.com/EOqwtNvJk6R8caV4O3fMt6kSjgM) (this is only with 10 elements btw).

Reply:
Never mind then.